waterfall/golang/server/server.go

```diff
@@ -262,19 +290,163 @@ func (s *WaterfallServer) Exec(rpc waterfall_grpc.Waterfall_ExecServer) error {
-			}
-			// the server always runs on android so the type assertion will aways succeed
-			panic("this never happens")
+	ec, err := exitCode(cmd.Wait())
+	if err != nil {
+		return err
+	}
+	return rpc.Send(&waterfall_grpc.CmdProgress{ExitCode: uint32(ec)})
+}
+
+type installReader struct{}
+
+func (r installReader) BuildMsg() interface{} {
+	return new(waterfall_grpc.InstallRequest)
+}
+
+// SetBytes sets the meessage bytes.
+func (r installReader) GetBytes(m interface{}) ([]byte, error) {
+	msg, ok := m.(*waterfall_grpc.InstallRequest)
+	if !ok {
+		// this never happens
+		panic("incorrect type")
+	}
+	return msg.Payload, nil
+}
+
+func shell(ctx context.Context, args []string) *exec.Cmd {
+	return exec.CommandContext(ctx, "/system/bin/sh", "-c", strings.Join(args, " "))
+}
+
+func getVersion() (int, error) {
+	cmd := exec.Command(getProp, sdkProp)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return 0, err
+	}
+	return strconv.Atoi(strings.Trim(string(out), "\r\n"))
+}
+
+// Installs installs the specified package on the device.
+func (s *WaterfallServer) Install(rpc waterfall_grpc.Waterfall_InstallServer) error {
+	ctx := rpc.Context()
+
+	ins, err := rpc.Recv()
+	if err != nil {
+		return err
+	}
+
+	// Ignore the error, well just default to legacy install
+	api, _ := getVersion()
+
+	// 3 Possible cases:
+	// Streaming installations are not available (api < 21): use pm install with temp file
+	// Streaming installations are available with pm (21 <= api < 24): use pm streamed install
+	// Streaming installations are available with cmd (api >= 24): use cmd streamed install
+	streamed := false
+	useCmd := false
+	if api >= 21 {
+		streamed = true
+	}
+	if api >= 24 {
+		useCmd = true
+	}
+
+	payload := stream.NewReader(rpc, installReader{})
+	if !streamed {
+		f, err := ioutil.TempFile(tmpDir, "*.apk")
+		if err != nil {
+			return err
+		}
+		defer os.Remove(f.Name())
+		defer f.Close()
+
+		if _, err := io.Copy(f, payload); err != nil {
+			return err
 		}
-		// we got some other kind of error
+
+		if err := f.Chmod(os.ModePerm); err != nil {
+			return err
+		}
+
+		cmd := shell(ctx, append([]string{pmCmd}, append(ins.Args, f.Name())...))
+		o, err := cmd.CombinedOutput()
+		s, err := exitCode(err)
+		if err != nil {
+			return err
+		}
+
+		return rpc.SendAndClose(
+			&waterfall_grpc.InstallResponse{
+				ExitCode: uint32(s),
+				Output:   string(o)})
+	}
+
+	insCmd := []string{pmCmd}
+	if useCmd {
+		insCmd = []string{cmdCmd, packageService}
+	}
+
+	action := append(insCmd, installCreate)
+	cmd := shell(ctx, append(action, ins.Args[1:]...))
+	o, err := cmd.CombinedOutput()
+	ec, err := exitCode(err)
+	if err != nil {
 		return err
 	}
-	return rpc.Send(&waterfall_grpc.CmdProgress{ExitCode: 0})
+
+	if ec != 0 {
+		return rpc.SendAndClose(
+			&waterfall_grpc.InstallResponse{
+				ExitCode: uint32(ec),
+				Output:   string(o)})
+	}
+
+	out := string(o)
+	ss := out[strings.Index(out, "[")+1 : strings.Index(out, "]")]
```

Comment:
If the session creation fails this will panic due to a bad index. It might be better to handle the error rather than crash the server.

Reply:
Im handling the error above. I added verification to the string just in case I get a bad output from pm.